src/charm.py

```diff
@@ -196,22 +207,45 @@ def __init__(self, *args):
             self.on.check_config_action, self._on_check_config  # pyright: ignore
         )
 
-    def set_ports(self):
-        """Open necessary (and close no longer needed) workload ports."""
-        planned_ports = {
-            OpenedPort("tcp", self._ports.api),
-            OpenedPort("tcp", self._ports.ha),
-        }
-        actual_ports = self.unit.opened_ports()
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_unit_status)
 
-        # Ports may change across an upgrade, so need to sync
-        ports_to_close = actual_ports.difference(planned_ports)
-        for p in ports_to_close:
-            self.unit.close_port(p.protocol, p.port)
+    def _on_collect_unit_status(self, event: CollectStatusEvent):
+        # Resource limit
+        if not self.resources_patch.is_ready():
+            event.add_status(
+                MaintenanceStatus("[resource patch] Waiting for resource limit patch to apply")
+            )
+        if self.resources_patch.last_error:
+            event.add_status(BlockedStatus(f"[resource patch] {self.last_error}"))
 
-        new_ports_to_open = planned_ports.difference(actual_ports)
-        for p in new_ports_to_open:
-            self.unit.open_port(p.protocol, p.port)
+        # Check open ports
+        if self.unit.opened_ports() != set(PORTS.values()):
+            event.add_status(WaitingStatus(f'Opening {", ".join(PORTS.keys())} ports'))
+
+        # "pebble ready"
+        if not self.container.can_connect():
+            event.add_status(
+                MaintenanceStatus(f"Waiting for '{self.container.name}' container to become ready")
+            )
+
+        # Make sure the external url is valid
+        if not _is_external_url_valid(self._external_url):
+            # This shouldn't happen
+            event.add_status(
+                BlockedStatus(
+                    f"Invalid external url: '{self._external_url}'; must include scheme and hostname."
+                )
+            )
+
+        # TLS
+        if self.server_cert.enabled and not self._is_tls_ready():
+            event.add_status(MaintenanceStatus("Waiting for TLS ready"))
+
+        if self.last_error:
+            event.add_status(BlockedStatus(self.last_error))
```

Comment:
I think this should also be error status I think because these, as far as I can tell, are miscellaneous un-handled errors.